rust/foxglove-derive/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+    };
+
+    let mut field_defs = Vec::new();
+    let mut field_encoders = Vec::new();
+    let mut enum_defs = Vec::new();
+    let mut message_defs = Vec::new();
+
+    for (i, field) in fields.iter().enumerate() {
+        let field_name = &field.ident.as_ref().unwrap();
+        let field_type = &field.ty;
+        let field_number = i as u32 + 1;
+
+        enum_defs.push(quote! {
+            if let Some(enum_desc) = <#field_type as ::foxglove::ProtobufField>::enum_descriptor() {
+                enum_type.push(enum_desc);
+            }
+        });
+
+        message_defs.push(quote! {
+            if let Some(message_descriptor) = <#field_type as ::foxglove::ProtobufField>::message_descriptor() {
+                nested_type.push(message_descriptor);
+            }
+        });
+
+        field_defs.push(quote! {
+            let mut field = foxglove::prost_types::FieldDescriptorProto::default();
+            field.name = Some(String::from(stringify!(#field_name)));
+            field.number = Some(#field_number as i32);
+
+            if <#field_type as ::foxglove::ProtobufField>::repeating() {
+                field.label = Some(foxglove::prost_types::field_descriptor_proto::Label::Repeated as i32);
+            } else {
+                field.label = Some(foxglove::prost_types::field_descriptor_proto::Label::Required as i32);
+            }
+            field.r#type = Some(<#field_type as ::foxglove::ProtobufField>::field_type() as i32);
+
+            field.type_name = <#field_type as ::foxglove::ProtobufField>::type_name();
+
+            message.field.push(field);
+        });
+
+        field_encoders.push(quote! {
+            ::foxglove::ProtobufField::write_tagged(&self.#field_name, #field_number, buf);
+        });
+    }
+
+    // Generate the output tokens
+    let expanded = quote! {
+        impl foxglove::ProtobufField for #name {
+            fn field_type() -> foxglove::prost_types::field_descriptor_proto::Type {
+                foxglove::prost_types::field_descriptor_proto::Type::Message
+            }
+
+            fn wire_type() -> u32 {
+                2 // Length-delimited, same as strings and bytes
+            }
+
+            fn write(&self, out: &mut impl foxglove::bytes::BufMut) {
+                use foxglove::bytes::BufMut;
+
+                let mut local_buf = vec![];
+
+                // make a mutable reference to buf because field_encoders needs a mutable reference
+                // for the generated code
+                let mut buf = &mut local_buf;
+
+                // Encode each field using proper protobuf encoding
+                #(#field_encoders)*
+
+                // Write the length as a varint
+                let len = buf.len();
+                let mut len_value = len as u64;
+                while len_value >= 0x80 {
+                    out.put_u8((len_value as u8) | 0x80);
+                    len_value >>= 7;
+                }
+                out.put_u8(len_value as u8);
```

Comment:
Could do something like this:

```suggestion
                let mut len_value = len as u64;
                ::foxglove::ProtobufField::write(&len_value, buf);
```

That's essentially a hack to get at the (inlined) `prost::encoding::encode_varint` that we want. Maybe we should re-export `foxglove::protobuf::encode_varint` under a `#[doc(hidden)]` and then we can just call that directly. Ditto for the enum variant encoding above.

Reply:
Done in #396; I re-exported `encode_varint` at the top level (as with `prost_file_descriptor_set_to_vec`).